### ISSUE TITLE
Encoding error while tracking something that contains Unicode characters using async

### ISSUE THREAD

**Issue by doitian** [getsentry/raven-ruby#725]
This is similar to issue https://github.com/getsentry/raven-ruby/issues/689 , which has been fixed by scrub event before sending to Sentry server.

However, the event is also serialized to JSON when `async` is enabled, and processors are not used here.

It is easy to reproduce it, just enable async and track rack request with unicode body

```ruby
require 'raven'

Raven.configure do |config|
  config.dsn = 'http://public:secret@example.com/project-id'
  config.async = lambda { |event|
    puts "ASYNC"
  }
end

use Raven::Rack
run Proc.new { |env|
  Raven.capture_message('request')
  ['200', {'Content-Type' => 'text/html'}, ['get rack\'d']]
}
```

Start the server using `rackup config.ru` and send the request

    curl -HContent-Type:text/javascript -XPOST -d'{"msg":"中文"}' http://localhost:9292

The server output:

```
I, [2017-08-02T10:57:06.594894 #78887]  INFO -- sentry: ** [Raven] Raven 2.6.1 ready to catch errors
[2017-08-02 10:57:06] INFO  WEBrick 1.3.1
[2017-08-02 10:57:06] INFO  ruby 2.2.3 (2015-08-18) [x86_64-darwin16]
[2017-08-02 10:57:06] INFO  WEBrick::HTTPServer#start: pid=78887 port=9292
E, [2017-08-02T10:57:07.807149 #78887] ERROR -- sentry: ** [Raven] async event sending failed: "\xE4" from ASCII-8BIT to UTF-8
```

Output of `bundle list`

```
bundle list
Gems included by the bundle:
  * bundler (1.14.4)
  * faraday (0.12.2)
  * multipart-post (2.0.0)
  * rack (2.0.3)
  * sentry-raven (2.6.1)
```

I have created a gist to contain all the files: https://gist.github.com/doitian/4af0abd56dd8ad4d3f8dcc4099f5e028

**Comment by doitian** [comment]
A workaround I use:

```ruby
require 'raven'                                                                                                                            
require 'raven/event'                                                                                                                      
                                                                                                                                           
module Raven                                                                                                                               
  class Event                                                                                                                              
    def to_json_compatible                                                                                                                 
      h = to_hash                                                                                                                          
      begin                                                                                                                                
        JSON.parse(JSON.generate(h))                                                                                                       
      rescue Encoding::UndefinedConversionError => e                                                                                       
        if h[:request] && h[:request][:data] then                                                                                          
          h[:request][:data].force_encoding(Encoding::UTF_8)                                                                               
          JSON.parse(JSON.generate(h))                                                                                                     
        else                                                                                                                               
          raise e                                                                                                                          
        end                                                                                                                                
      end                                                                                                                                  
    end                                                                                                                                    
  end                                                                                                                                      
end
```

**Comment by nateberkopec** [comment]
Thanks for the repro 👍

**Comment by nateberkopec** [comment]
We started converting the event to a JSON-compatible hash in a786def7b423 / v2.0.0.

**Comment by nateberkopec** [comment]
Hmmm, maybe we should just run the processor chain before providing the event to `async`, and then not run the processor chain again when send_event is called.

**Comment by nateberkopec** [comment]
That's a major change though. In the meantime, I'm going to run the event through the UTF-8 and circular reference processors. It will run through them twice (once while creating the hash to give to `async`, once while sending) but anything else would be a major API change.

**Comment by nateberkopec** [comment]
Closed by #730
